File: deploy/tests/polkit_rule_test.js
"use strict";

const fs = require("fs");
const vm = require("vm");

let rule;
const sandbox = {
    polkit: {
        Result: { YES: "YES", NOT_HANDLED: "NOT_HANDLED" },
        addRule: candidate => { rule = candidate; },
    },
};
const source = fs.readFileSync(process.argv[2], "utf8");
vm.runInNewContext(source, sandbox, { filename: process.argv[2] });
if (typeof rule !== "function") throw new Error("polkit rule was not registered");

function action(id, unit, verb) {
    return { id, lookup: key => ({ unit, verb })[key] };
}
function expect(label, got, want) {
    if (got !== want) throw new Error(`${label}: got ${got}, want ${want}`);
}

const yes = sandbox.polkit.Result.YES;
const no = sandbox.polkit.Result.NOT_HANDLED;
const agent = { user: "pgw-agent" };
expect("valid start", rule(action("org.freedesktop.systemd1.manage-units", "pgw-fwd@15001.service", "start"), agent), yes);
expect("valid stop", rule(action("org.freedesktop.systemd1.manage-units", "pgw-fwd@15999.service", "stop"), agent), yes);
expect("different user", rule(action("org.freedesktop.systemd1.manage-units", "pgw-fwd@15001.service", "start"), { user: "pgw-api" }), no);
expect("wrong action", rule(action("org.freedesktop.systemd1.manage-unit-files", "pgw-fwd@15001.service", "start"), agent), no);
expect("status verb", rule(action("org.freedesktop.systemd1.manage-units", "pgw-fwd@15001.service", "status"), agent), no);
expect("other unit", rule(action("org.freedesktop.systemd1.manage-units", "pgw-api.service", "restart"), agent), no);
expect("template unit", rule(action("org.freedesktop.systemd1.manage-units", "pgw-fwd@.service", "restart"), agent), no);
expect("leading zero", rule(action("org.freedesktop.systemd1.manage-units", "pgw-fwd@015001.service", "restart"), agent), no);
expect("below range", rule(action("org.freedesktop.systemd1.manage-units", "pgw-fwd@15000.service", "restart"), agent), no);
expect("above range", rule(action("org.freedesktop.systemd1.manage-units", "pgw-fwd@16000.service", "restart"), agent), no);

process.stdout.write("polkit rule tests: PASS\n");

File: deploy/tests/polkit_rule_test.py
#!/usr/bin/python3
"""Byte-pinned structural lint only; JavaScript semantics are tested by Node."""

import hashlib
import sys


PINNED_SHA256 = "36304260b626f196a4e3eecbe8480c3b0857121cdcec4172416e0b6f4dfa399c"


def deny_unless(condition, message):
    if not condition:
        raise SystemExit(message)


def main():
    if len(sys.argv) != 2:
        raise SystemExit("usage: polkit_rule_test.py RULE")
    raw = open(sys.argv[1], "rb").read()
    deny_unless(b"\x00" not in raw and b"\r" not in raw, "unsafe rule encoding")
    deny_unless(hashlib.sha256(raw).hexdigest() == PINNED_SHA256,
                "polkit rule bytes differ from the reviewed structural pin")
    source = raw.decode("utf-8")
    required = (
        'action.id !== "org.freedesktop.systemd1.manage-units"',
        'subject.user !== "pgw-agent"',
        '["start", "stop", "restart"].indexOf(verb) === -1',
        r'/^pgw-fwd@([0-9]+)\.service$/.exec(unit)',
        "port < 15001 || port > 15999 || String(port) !== match[1]",
    )
    for snippet in required:
        deny_unless(snippet in source, "polkit rule missing exact guard: " + snippet)
    deny_unless(source.count("return polkit.Result.YES;") == 1, "polkit rule has multiple allow exits")
    deny_unless(source.count("polkit.addRule(") == 1, "polkit rule count is not exact")
    print("polkit byte-pinned structural lint: PASS")


if __name__ == "__main__":
    main()
